### PR TITLE
Added Dockerfile and docker-compose.yml

### PR DESCRIPTION
**Pull request by chris-alexiuk**
Extension of #91

### CODE REVIEW

**Comment by tloen** [comment]
Have been running into some problems testing this and #91 on my local machine, apologies for the delay

**Comment by chris-alexiuk** [comment]
Anything I can address?

**Comment by tloen** [comment]
Let me make sure it's not on my side first. I've been seeing

```
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
```

which I think has to do with my own nvidia docker setup.

**Comment by tloen** [comment]
Alright — it works, although inference is a bit slower (possibly because of CUDA version/bitsandbytes/4090). I'll do some polishing, add @FrancescoSaverioZuppichini as a coauthor, and we can ship this. Thanks for the work!

**Comment by chris-alexiuk** [comment]
Thanks, glad you got it working!

**Comment by FrancescoSaverioZuppichini** [comment]
@tloen GitHub (above) says this branch was merged in main but I don't see it, I've left a comment on the code

File: README.md [hunk 1]
```diff
@@ -94,6 +82,49 @@ They should help users
...
+1. Build the container image:
+
+```bash
+docker build -t alpaca-lora .
+```
+
+2. Run the container (you can also use `finetune.py` and all of its parameters as shown above for training):
+
+```bash
+docker run --gpus=all --shm-size 64g -p 7860:7860 -v ${HOME}/.cache:/root/.cache --rm alpaca-lora generate.py \
```
Comment (FrancescoSaverioZuppichini) [inline README.md:96]: We should also add `--ipc host`

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  alpaca-lora:
+    build:
+      context: ./
+      dockerfile: Dockerfile
+      args:
+        BUILDKIT_INLINE_CACHE: "0"
+    image: alpaca-lora
+    shm_size: '64gb'
+    command: generate.py --load_8bit --base_model $BASE_MODEL --lora_weights 'tloen/alpaca-lora-7b'
+    restart: unless-stopped
```
Comment (FrancescoSaverioZuppichini) [inline docker-compose.yml:13]: we should also add `ipc: host`

**Review by FrancescoSaverioZuppichini — COMMENTED** [review]
left a minor comment

**Comment by chris-alexiuk** [comment]
I would prefer to not use `ipc: host` as it's a slightly less safe way of exposing the network - I've modified the gradio `.launch()` to include `server_name="0.0.0.0"` which will allow the container to be accessed from outside through port `7860`.

**Comment by FrancescoSaverioZuppichini** [comment]
@chris-alexiuk but I think that is required when using multi processes (at least this is what I remember),  I don't think the network is an issue since one port is shared to the host and only that one

**Pull request merged**